analyses/comparative-RNASeq-analysis/02-thresholds-and-outliers.py

```diff
@@ -0,0 +1,231 @@
+    iqr = quartile3 - quartile1
+    outlier_range = iqr_multiplier * iqr
+    return (quartile3 + outlier_range, quartile1 - outlier_range)
+
+def modified_tukey_thresholds(expression_values, iqr_multiplier):
+    """Use a modified Tukey method to generate outlier thresholds (high, low)
+    for a single gene. Drops the highest value when calculating the high threshold,
+    and the lowest value for low threshold, to match thresholds for outlier
+    status when a sample is not present in the original dataset."""
+    drop_highest = expression_values.drop(expression_values.idxmax())
+    high = tukey_thresholds(drop_highest, iqr_multiplier)[0]
+    drop_lowest = expression_values.drop(expression_values.idxmin())
+    low = tukey_thresholds(drop_lowest, iqr_multiplier)[1]
+    return( pd.Series([high, low], index=["high", "low"]))
+
+def generate_all_thresholds(normalized_samples,
+                        scratch_dir,
+                        iqr_multiplier=1.5,
+                        prefix="",
+                        verbose=False):
+    """Generate high and low outlier thresholds for each gene
+    using the modified tukey method from a matrix of
+    log2(TMP+1) normalized sample expression"""
+    print_v = print if verbose else lambda *a, **k: None
+
+    print_v("Calculating outlier thresholds.")
+    thresholds = normalized_samples.apply(modified_tukey_thresholds, args=[iqr_multiplier], axis=1)
+    thresholds_path = os.path.join(scratch_dir, "{}threshold-expression-values.rds".format(prefix))
+    print_v("Writing expression outlier thresholds to {}".format(thresholds_path))
+    utils.write_rds(thresholds, thresholds_path)
+    return thresholds
+
+def single_sample_outliers(sample_expression, expression_thresholds):
+    """Given the expression of a single sample as series, and matrix
+    of thresholds, calculate the up and down outliers and genes in the top 5%
+    of sample's expression. Return a matrix of genes vs samples, with each gene
+    marked as U or D for outlier, T if it's in the top 5%, and F for qc Fail."""
+
+    down_outliers = sample_expression[sample_expression < expression_thresholds["low"]].index
+    up_outliers = sample_expression[sample_expression > expression_thresholds["high"]].index
+
+    # Mark genes with expression in the top 5% within that sample (inclusive)
+    top5_value = sample_expression.quantile(0.95)
+    qc_fail = (top5_value <=0) # If 0 is in the top5%, this sample is really underexpressed
+    top5_genes = sample_expression[sample_expression >= top5_value].keys()
+
+    # Possible values include (D)own outlier, (U)p outlier, (T)op5%, and qc (F)ail
+    # dro(P)ped genes are added in calculate_outliers
+    result = pd.Series("", index = sample_expression.index)
+    result[top5_genes] += "T"
+    result[down_outliers] += "D"
+    result[up_outliers] += "U"
+    if qc_fail:
+        result += "F" # Mark every gene as QC fail for this sample
```

Comment:
Reading a bit further, I see that you would need one more character for the dropped/retained code, maybe D/R (if position matters, redundant letters are available, but maybe we would still want to avoid them?)